src/boot/efi/boot.c

```diff
@@ -1951,6 +1961,95 @@ static VOID config_entry_add_linux(
         uefi_call_wrapper(linux_dir->Close, 1, linux_dir);
 }
 
+#ifndef DEVICE_TREE_GUID
+#define DEVICE_TREE_GUID \
+        { 0xb1b621d5, 0xf19c, 0x41a5, {0x83, 0x0b, 0xd9, 0x15, 0x2c, 0x69, 0xaa, 0xe0} }
+#endif
+static const EFI_GUID fdt_guid = DEVICE_TREE_GUID;
+
+static EFI_STATUS devicetree_get(const void **fdt) {
+        EFI_CONFIGURATION_TABLE *entry = ST->ConfigurationTable;
+        EFI_CONFIGURATION_TABLE *end = entry + ST->NumberOfTableEntries;
+
+        for (; entry < end; entry++) {
+                if (CompareGuid(&entry->VendorGuid, (EFI_GUID *)&fdt_guid) == 0) {
+                        *fdt = entry->VendorTable;
+                        return EFI_SUCCESS;
+                }
+        }
+
+        return EFI_UNSUPPORTED;
+}
+
+static const void *devicetree_orig;
+static EFI_PHYSICAL_ADDRESS devicetree_addr;
+static UINTN devicetree_pages;
```

Comment:
hmm, wa? we try to avoid global state. why this?

Reply:
I'm storing this information between devicetree_install() and devicetree_restore(), so we can restore the devicetree if we fail to start the kernel. This is technically not needed, but I thought it would be nice for whatever comes after sd-boot that it doesn't get the devicetree that was specified in the .conf that failed to boot.

Alternatively we could store this on the stack of the caller, start_image. Would that be better?

Reply:
maybe add a struct SavedDeviceTreeData or so, and put that on the stack in image_start() and pass a pointer to both devicetree_install() and devicetree_restore().

We try to avoid global state with some rare exceptions. Stuff that isn't inherently per-process (such as let's say variables containing the parsed process' cmdline args) really shouldn't be global/static variables.

Reply:
Yes, that's exactly the alternative I had in mind. I'll do that.